### PR TITLE
Add Scaleway builder

### PR DESCRIPTION
**Pull request by edouardb**
Add [Scaleway](https://www.scaleway.com/) builder for Packer.

### CODE REVIEW

File: builder/scaleway/config.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	var errs *packer.MultiError
+	if es := c.Comm.Prepare(&c.ctx); len(es) > 0 {
+		errs = packer.MultiErrorAppend(errs, es...)
+	}
+	if c.Organization == "" {
+		errs = packer.MultiErrorAppend(
+			errs, errors.New("Scaleway Organization ID must be specified"))
+	}
+
+	if c.Token == "" {
```
Comment (edouardb) [inline builder/scaleway/config.go]: done

**Comment by pdutourgeerling** [comment]
I just tried the Scaleway builder, for the most part it works great but there were several things I came across.

The most important issue is that not everything is written to the disk when there is no wait time between the last provisioner and stopping the server.

To make it better understandable I created two test templates:

### `test1`
```json
{
  "builders": [{
    "type": "scaleway",
    "api_organization": "-",
    "api_token": "-",
    "image": "7069f3d0-23e3-4664-8f6b-1a17ed478dff",
    "region": "par1",
    "commercial_type": "VC1S",
    "ssh_username": "root",
    "snapshot_name": "Docker test 1 {{timestamp}}"
  }],
  "provisioners": [{
    "type": "shell",
    "inline": [
      "yum install -y yum-utils",
      "yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo",
      "yum install docker-ce -y",
      "cat /usr/lib/systemd/system/docker.service"
    ]
  }]
}
```

### `test2`
```json
{
  "builders": [{
    "type": "scaleway",
    "api_organization": "-",
    "api_token": "-",
    "image": "7069f3d0-23e3-4664-8f6b-1a17ed478dff",
    "region": "par1",
    "commercial_type": "VC1S",
    "ssh_username": "root",
    "snapshot_name": "Docker test 2 {{timestamp}}"
  }],
  "provisioners": [{
    "type": "shell",
    "inline": [
      "yum install -y yum-utils",
      "yum-config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo",
      "yum install docker-ce -y",
      "cat /usr/lib/systemd/system/docker.service"
    ]
  }, {
    "type": "shell",
    "inline": [
      "sleep 300"
    ]
  }]
}
```

When Packer is building and shows the content of the `docker.service` file, it shows the right content on both `test1` and `test2`. Once I start new servers based on the created snapshots and show the content of the `docker.service` file the file is empty on the server based on the `test1` snapshot and the file contains the right content on the server based on the `test2` snapshot.

While I haven't tried it without Packer I think this issue has more to do with Scaleway then with the Scaleway Packer builder plugin.

Here are some less important things I came across while testing:
- The documentation doesn't say how you can find your organization ID, after some API calls I found out that the organization ID is called 'access key' on the Scaleway website.
- It isn't possible to rename the created volume. Created volumes have the same name as the base image which makes it difficult to know which volume is associated with a snapshot.
- It isn't possible to let Packer create an image from the snapshot.
- `-on-error=cleanup` doesn't work, it always exits without any cleanup

**Comment by edouardb** [comment]
@rickard-von-essen is everything ok?

**Comment by rickard-von-essen** [comment]
@edouardb Sorry had not time to really test this yet. But at least managed to register now.

File: builder/scaleway/ssh.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+func sshConfig(state multistep.StateBag) (*ssh.ClientConfig, error) {
+	config := state.Get("config").(Config)
+	privateKey := state.Get("privateKey").(string)
+
+	signer, err := ssh.ParsePrivateKey([]byte(privateKey))
+	if err != nil {
+		return nil, fmt.Errorf("Error setting up SSH config: %s", err)
+	}
+
+	return &ssh.ClientConfig{
```
Comment (edouardb) [inline builder/scaleway/ssh.go]: Just added support for existing ssh key and updated the doc

**Comment by pdutourgeerling** [comment]
@edouardb Any idea what caused the 'disappearing' data issue? Do you think it has something to do with the builder or Scaleway itself?

**Comment by edouardb** [comment]
@pdutourgeerling Not related to the builder, I tried to reproduce and everything was fine.

**Comment by edouardb** [comment]
Hi @rickard-von-essen any news?

**Comment by rickard-von-essen** [comment]
Sorry, I'm very occupied at the moment and won't be able to do much Packer work the coming 4 weeks at least.

**Comment by edouardb** [comment]
@rickard-von-essen are you back :) ?

**Comment by rickard-von-essen** [comment]
Sorry still very busy, and Matt too. Maybe @SwampDragons has sometime to dedicate to this?

**Comment by SwampDragons** [comment]
I'm focused on 1.0.3 but I'll try to take a look when I get a chance.

File: website/source/docs/builders/scaleway.html.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+### Optional:
+
+- `server_name` (string) - The name assigned to the server.
+
+- `image_name` (string) - The name of the resulting image that will
+    appear in your account.
+
+- `snapshot_name` (string) - The name of the resulting snapshot that will
+    appear in your account.
```
Comment (rickard-von-essen) [inline website/source/docs/builders/scaleway.html.md]: Add info about default values.

File: builder/scaleway/ssh.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+		User: config.Comm.SSHUsername,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+	}, nil
+}
```
Comment (rickard-von-essen) [inline builder/scaleway/ssh.go]: I would like `ssh_agent_auth` and `ssh_password` support here too.

File: builder/scaleway/builder.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+	"github.com/hashicorp/packer/common"
+	"github.com/hashicorp/packer/helper/communicator"
+	"github.com/hashicorp/packer/packer"
+	"github.com/mitchellh/multistep"
+	"github.com/scaleway/scaleway-cli/pkg/api"
+)
+
+// The unique id for the builder
+const BuilderId = "pearkes.scaleway"
```
Comment (rickard-von-essen) [inline builder/scaleway/builder.go]: Copy-paste? change id to `hashicorp.scaleway`

**Comment by rickard-von-essen** [comment]
And when I tried to run the example included in the documentation I get this:

```
scaleway output will be in this color.

==> scaleway: Using existing SSH private key
==> scaleway: Creating server...
==> scaleway: Error creating server: StatusCode: 405, Type: 405, APIMessage: The method is not allowed for the requested URL.
==> scaleway: Destroying server...
==> scaleway: Error destroying server. Please destroy it manually: StatusCode: 405, Type: 405, APIMessage: The method is not allowed for the requested URL.
Build 'scaleway' errored: Error creating server: StatusCode: 405, Type: 405, APIMessage: The method is not allowed for the requested URL.

==> Some builds didn't complete successfully and had errors:
--> scaleway: Error creating server: StatusCode: 405, Type: 405, APIMessage: The method is not allowed for the requested URL.

==> Builds finished but no artifacts were created.
```

File: builder/scaleway/ssh.go (hunk 4 above)
Comment (edouardb) [inline builder/scaleway/ssh.go]: Scaleway only provides auth via ssh key at the moment.

**Comment by edouardb** [comment]
@rickard-von-essen can you show me your config, just try and didn't encounter any issue.

File: builder/scaleway/ssh.go (hunk 4 above)
Comment (rickard-von-essen) [inline builder/scaleway/ssh.go]: But I expect there is nothing stopping people from enabling password auth and setting some passwords. Especially when they have tool for creating their own images.

**Comment by rickard-von-essen** [comment]
@edouardb basically the example from the docs.

```json
{
  "builders": [
    {
      "type": "scaleway",
      "api_access_key": "",
      "api_token": "",
      "image": "f01f8a48-c026-48ac-9771-a70eaac0890e",
      "region": "par1",
      "commercial_type": "X64-2GB",
      "ssh_username": "root",
      "ssh_private_key_file": "id_rsa-scaleway-npw"
    }
  ]
}
```

File: website/source/docs/builders/scaleway.html.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+{
+  "type": "scaleway",
+  "api_access_key": "YOUR API ACCESS KEY",
+  "api_token": "YOUR TOKEN",
+  "image": "UUID OF THE BASE IMAGE",
+  "region": "par1",
+  "commercial_type": "X64-2GB",
+  "ssh_username": "root",
+  "ssh_private_key_file": "~/.ssh/id_rsa",
+  Extra,
```
Comment (rickard-von-essen) [inline website/source/docs/builders/scaleway.html.md]: What I ment was that you have an extra comma on the line above.  😉

File: website/source/docs/builders/scaleway.html.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+### Optional:
+
+- `server_name` (string) - The name assigned to the server. Default `packer-UUID`
+
+- `image_name` (string) - The name of the resulting image that will
+    appear in your account. Default `packer-TIMESTAMP`
+
+- `snapshot_name` (string) - The name of the resulting snapshot that will
+    appear in your account. Default `packer-TIMESTAMP
```
Comment (rickard-von-essen) [inline website/source/docs/builders/scaleway.html.md]: Missin closing `

File: website/source/docs/builders/scaleway.html.md (hunk 6 above)
Comment (edouardb) [inline website/source/docs/builders/scaleway.html.md]: OMG, I'll fix

**Comment by edouardb** [comment]
The image UUID you specified in your config is not yet valid. I upgraded the documentation with details on how to get the image UUID

**Comment by rickard-von-essen** [comment]
>  I upgraded the documentation with details on how to get the image UUID

Great!

**Comment by rickard-von-essen** [comment]
> The image UUID you specified in your config is not yet valid. 

Ok

> Error creating server: StatusCode: 405, Type: 405, APIMessage: The method is not allowed for the requested URL.

This is a really bad error message. In my world it says _"you are using GET instead of POST or vice-versa"_

It should state _what_ is wrong with the request.

**Comment by edouardb** [comment]
The 405 error code is not related to the create server request but is due to a missing error handling in the builder. Just fixed it.

**Comment by rickard-von-essen** [comment]
```json
{
  "builders": [
    {
      "type": "scaleway",
      "api_access_key": "",
      "api_token": "",
      "image": "857de28c-39ce-4b15-9389-b68d8bb1bb51",
      "region": "par1",
      "commercial_type": "X64-2GB",
      "ssh_username": "root",
      "ssh_private_key_file": "id_rsa-scaleway-npw"
    }
  ]
}
```

```
scaleway output will be in this color.

==> scaleway: Using existing SSH private key
==> scaleway: Creating server...
==> scaleway: Waiting for server to become active...
==> scaleway: Waiting for SSH to become available...
==> scaleway: Connected to SSH!
==> scaleway: Shutting down server...
==> scaleway: Creating snapshot: snapshot-packer-1500530755
==> scaleway: Creating image: image-packer-1500530755
==> scaleway: Terminating server...
==> scaleway: Destroying server...
==> scaleway: Error destroying server. Please destroy it manually: StatusCode: 400, Type: invalid_request_error, APIMessage: server should be running
Build 'scaleway' finished.

==> Builds finished. The artifacts of successful builds are:
--> scaleway: An image was created: 'image-packer-1500530755' (ID: 4a9066e2-a4de-45fe-8b99-a05520f413b1) in region 'par1' based on snapshot 'snapshot-packer-1500530755' (ID: c15f00f5-ba87-4415-a3a9-8dd0c0461f49)
```

**Comment by rickard-von-essen** [comment]
Setting `"region": "par2"` makes it panic.

**Comment by rickard-von-essen** [comment]
And maybe it would be nice to set the size of the boot volume if posible and some tags etc.

**Comment by edouardb** [comment]
@rickard-von-essen it's not possible to set the size of the boot volume.

**Comment by rickard-von-essen** [comment]
> it's not possible to set the size of the boot volume.

Ok

**Comment by edouardb** [comment]
@rickard-von-essen All good ?

File: builder/scaleway/builder.go [hunk 8]
```diff
@@ -33,7 +33,11 @@ func (b *Builder) Prepare(raws ...interface{}) ([]string, error) {
 }
 
 func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
-	client, _ := api.NewScalewayAPI(b.config.Organization, b.config.Token, b.config.UserAgent, b.config.Region)
+	client, err := api.NewScalewayAPI(b.config.Organization, b.config.Token, b.config.UserAgent, b.config.Region)
+
+	if err != nil {
+		return nil, err
```
Comment (rickard-von-essen) [inline builder/scaleway/builder.go]: Add `ui.Error(err.Error())` before this.

**Comment by rickard-von-essen** [comment]
And a rebase...

**Comment by rickard-von-essen** [comment]
And if I understand it correctly the volume after the instance should be cleaned up, if it's not needed.

File: builder/scaleway/builder.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+			Host:      commHost,
+			SSHConfig: sshConfig,
+		},
+		new(common.StepProvision),
+		new(stepShutdown),
+		new(stepSnapshot),
+		new(stepImage),
+	}
+
+	b.runner = common.NewRunner(steps, b.config.PackerConfig, ui)
```
Comment (rickard-von-essen) [inline builder/scaleway/builder.go]: This is the wrong way of doing it. Check how it's done in [virtualbox-iso](https://github.com/hashicorp/packer/blob/master/builder/virtualbox/iso/builder.go#L296).

File: builder/scaleway/builder.go (hunk 8 above)
Comment (dimtion) [reply builder/scaleway/builder.go]: Fixed in 2cc13e74fb650d31f3fa64d2c23b1baf3cdeec7a

**Comment by dimtion** [comment]
Hi @rickard-von-essen ,

I've rebased the branch to avoid any conflicts in the vendor file, I've also added support for the password auth with the commit 7890d6028234ed76a268ab9d61747a5c8cac4c68, fixed the missing ui report you have highlighted in 2cc13e74fb650d31f3fa64d2c23b1baf3cdeec7a and made use of the correct runner creation method in a60bfb37d7ec847b23009ad0f0b563c8dfb41987.

It seems that the builds are not starting due to dependencies errors, I'm on to it.

 If you want me to squash some commits, or rebase some stuff please tell me.

**Comment by rickard-von-essen** [comment]
@dimtion great after the fix of the deps I think it's only `ssh_agent_auth` left to add and this seems ready.

**Comment by dimtion** [comment]
It seems like you need to add the package `github.com/scaleway/scaleway-cli/pkg/api` to your builder in order to make the tests pass.

**Comment by rickard-von-essen** [comment]
@mwhooker & @SwampDragons signing this over to you to merge when appropriate or do any more testing/reviewing you see fit.

**Comment by SwampDragons** [comment]
We're prepping for the 1.2.0 release next week -- can I get a clean rebase of this to review and merge?

**Comment by mwhooker** [comment]
I fixed some merge conflicts. Will need to test this ourselves before merging, which generally means running a json config. I think I have access to scaleway credentials, but if not I may ask you for some temporary ones.

We're planning to release this coming Friday, and planning to get this in.

**Comment by mwhooker** [comment]
works beautifully, thanks all!

**Comment by mwhooker** [comment]
@edouardb @dimtion Could you please email me at mhooker <at> hashicorp.com? I'd like to discuss who's going to maintain this. I'd like to add you to the contributors file and give you commit bit

**Pull request merged**